### DISCUSSION TITLE
SyncFusion Grid

### DISCUSSION THREAD

**Discussion by kcabral817** [MarimerLLC/csla#3217] [Questions]
Is anyone else using the Syncfusion Blazor grid?  I'm having a devil of a time getting the edit and add new to work correctly.  I hope this isn't too vague, but I'm looking to see if anyone else has overcome the automatic instantiation of a business object.

**Comment by rockfordlhotka** [comment]
Hopefully the grid raises an event when a new item is added/created? If so, you can probably use that event to get access to the newly created instance, use the `IUseApplicationContext` interface on the business object, and set its `ApplicationContext` property. That should get the object to be functional.

...

After I typed that, it occurs to me that the grid probably creates the instance, then adds it to the business collection. The business collection _does_ have a method you can override when a new item is added, and you could set `ApplicationContext` via the `IUseApplicationContext` interface at that point.

**Comment by kcabral817** [comment]
The grid is instantiating the class calling the parameter less constructor.  When it does this, the ID of the object is null and that throws an error.  I've been trying to intercept the invocation and generation.  Your suggestion caught me off guard as I thought the application context was for principles, auth and authenticated.  I am reading chapter 5 again to see how this would work :-).

**Reply by rockfordlhotka** [reply]
In CSLA 6 the `ApplicationContext` needs to be set as the object is created. That's not true for previous versions.

When you say the "id" are you talking about your object id property, or the private unique identity value of the object within the object graph?

CSLA creates all domain objects using the default constructor too. Before CSLA 6 it was `Activator.CreateInstance`, and in CSLA 6 it is that plus using `IUseApplicationContext` to set the current scoped context.

**Comment by kcabral817** [comment]
I think the Id issue is a red herring.  It's the private unique field of the instance from the db.  

Here is what my class looks like:  

```
[Serializable]
    [Csla.Server.ObjectFactory(typeof(ChartOfAccountFactory), "Create", "GetAsync", "UpdateAsync", "DeleteAsync")]
    public class ChartOfAccountInfoPersist : BusinessBase<ChartOfAccountInfoPersist>, IUseApplicationContext
    {
        public ChartOfAccountInfoPersist()
        {
            MarkAsChild();
            MarkOld();  //Set new in the factory create method.
        }
```

(I append my business classes with Info if read only.  I append InfoPersist when it's editable.  The base class ChartOfAccount, tells me it's the class from the database using EF Core.  When I'm looking through the system I can instantly tell where in the architecture the class is located).  These types of objects are contained in lists--such things as code lists, chart of accounts, and typically two or three attributes for the class I use in lists for ease of editing and viewing.  That's how they are edited so they have the MarkAsChild() in the constructor.  Moreover, because I'm using Automapper and use graphs a lot, I needed to include MarkAsChild() in the default constructor.  If I remove it, I have no idea how to have Automapper map a graph structure.  If you think this is a bad idea, let me know.  It works well with WPF under the older .Net framework, but these are new times.

I just started adding the IUseApplicationContext interface based on your suggestion.  Do I need to do anything else with the class so that it is initialized?

**Reply by rockfordlhotka** [reply]
The only thing CSLA requires is that `ApplicationContext` be set immediately after the object is created. When the object gets added to the parent list then some other things happen, but that's all automatic.

Marking "new" objects as old isn't a bad thing as long as you understand the semantic difference and how it affects the data portal calling update or insert. It is relatively common to make this adjustment in the framework depending on how folks want a "new" object to be treated.

I suspect the big thing for you is that your object isn't initialized via a Create operation, so you don't get any id or whatever from the database. Perhaps the solution is to set the id value to an impossible value (-1 or something) so you can detect that this object was created and has no key from the database yet?
